app/src/main/java/com/jetbrains/handson/mpp/mobile/MainActivity.kt

```diff
@@ -28,18 +27,28 @@ class MainActivity : AppCompatActivity(), ApplicationContract.View {
     }
 
     override fun showResults(result: List<JourneyTableDataElem>) {
-        findViewById<TextView>(R.id.search_results).text = result.toString()
+        findViewById<RecyclerView>(R.id.search_results).adapter = SearchResultAdapter(result)//.text = result.toString()
     }
 
-    override fun populateStationList(stations: List<String>) {
-        findViewById<Spinner>(R.id.from_station)?.adapter =
+    override fun populateStationList(stations: List<JourneyStation>) {
+        departureSpinner = findViewById(R.id.from_station)
+        arrivalSpinner = findViewById(R.id.to_station)
+        departureSpinner?.adapter =
             ArrayAdapter(this, android.R.layout.simple_spinner_dropdown_item, stations)
-        findViewById<Spinner>(R.id.to_station)?.adapter =
+        arrivalSpinner?.adapter =
             ArrayAdapter(this, android.R.layout.simple_spinner_dropdown_item, stations)
     }
 
     private fun setupSearchListener(presenter: ApplicationPresenter) {
         val button: Button = findViewById(R.id.search_button)
-        button.setOnClickListener { presenter.onSearchClicked(timeUTCString = null) }
+        button.setOnClickListener {
+            val fromStation = departureSpinner?.selectedItem as JourneyStation?
+            val toStation = arrivalSpinner?.selectedItem as JourneyStation?
+            if(fromStation != null && toStation!=null){
+                presenter.onSearchClicked(fromStation.crsCode,toStation.crsCode,timeUTCString = null)
```

Comment:
Nit: spacing between item here and in line above. Also excess new line spacing below